zeus21/maps.py

```diff
@@ -108,18 +112,394 @@ def __init__(self, T21_coefficients, Power_Spectrum, z, Lbox=600, Nbox=200, KIND
+        self.z = np.atleast_1d(input_z) #CoeffStructure.zintegral[self._z_idx]
+
+        ### generating the density field at the closest redshift to the lower one inputed
+        self.z_of_density = self.z[0]
+        self.density = self.generate_density(ClassyCosmo, CorrFClass)
+        self.density_allz = np.empty((len(self.z), self.ncells, self.ncells, self.ncells), dtype=np.float32)
+        if self.COMPUTE_DENSITY_AT_ALLZ:
+            self.generate_density_allz(CosmoParams)
+
+        ### smoothing the density field
+        self._k = self.compute_k()
+        self.density_smoothed_allr = self.smooth_density()
+
+        ### generating the ionized field, and computing the ionized fraction
+        self.barrier = barrier
+        if self.barrier is None:
+            self.barrier = np.array([BMF.B(z, self.r) for z in self.z]) #BMF linear barrier
+        self.ion_field_allz, self.ion_frac = self.generate_xHII(CosmoParams, CoeffStructure, BMF)
+
+        ### computing the mass weighted ionized fraction
+
+        self._has_mw = False
+        self.lowres_massweighting = lowres_massweighting
+        if self.COMPUTE_MASSWEIGHTED:
+            self.compute_massweighted(CosmoParams, self.lowres_massweighting)
+
+        self._has_p = False
+        if self.COMPUTE_PARTIAL_IONIZATIONS:
+            self.compute_partial(CosmoParams, BMF)
+
+        self._has_mwp = False
+        if self.COMPUTE_PARTIAL_AND_MASSWEIGHTED:
+            self.compute_partial_massweighted(CosmoParams, BMF)
+
+        if self.COMPUTE_ZREION:
+            self.zreion = self.compute_zreion_frombinaryxHII()
+            self.treion = self.compute_treion(ClassyCosmo)
+            
+        
+        if self.PRINT_TIMER:
+            z21_utilities.print_timer(self._start_time, text_before="Total computation time: ")
+        
+
+    def generate_density(self, ClassyCosmo, CorrFClass):
+        if self.PRINT_TIMER:
+            start_time = time.time()
+            print("Generating density field...")
+        #Generating matter power spectrum at the lowest redshift
+        klist = CorrFClass._klistCF
+        pk_matter = np.zeros_like(klist)
+        for i, k in enumerate(klist):
+            pk_matter[i] = ClassyCosmo.pk(k, self.z_of_density)
+        pk_spl = spline(np.log(klist), np.log(pk_matter))
+    
+        #generating density map
+        if self.LOGNORMAL_DENSITY:
+            pb = pbox.LogNormalPowerBox(N=self.ncells, dim=3, pk=(lambda k: np.exp(pk_spl(np.log(k)))), boxlength=self.boxlength, seed=self.seed)
+        else:
+            pb = pbox.PowerBox(N=self.ncells, dim=3, pk=(lambda k: np.exp(pk_spl(np.log(k)))), boxlength=self.boxlength, seed=self.seed)
+        density_field = pb.delta_x().astype(np.float32, copy=False)
+        if self.PRINT_TIMER:
+            z21_utilities.print_timer(start_time, text_before="    done in ")
+        return density_field
+
+    def generate_density_allz(self, CosmoParams):
```

Comment:
are you generating a delta box at each z? why? you can just use the same box and multiply by growth on the fly no?